src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -179,17 +189,37 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
       // should not be here, and so ReserveCostValue should not be used
       log.warn("Cost is empty in processMempoolResult")
     }
+
     val cost = costOpt.getOrElse(FallbackCostValue)
-    val ng = processingResult match {
+
+    val newInterblockCost = processingResult match {
       case _: FailedTransaction => interblockCost.copy(invalidatedCost = interblockCost.invalidatedCost + cost)
       case _: SuccessfulTransaction => interblockCost.copy(acceptedCost = interblockCost.acceptedCost + cost)
       case _: DeclinedTransaction => interblockCost.copy(declinedCost = interblockCost.declinedCost + cost)
     }
 
-    log.debug(s"Old global cost info: $interblockCost, new $ng, tx processing cache size: ${txProcessingCache.size}")
-    interblockCost = ng
+    log.debug(s"Old global cost info: $interblockCost, " +
+      s"new: $newInterblockCost, tx processing cache size: ${txProcessingCache.size}")
+    interblockCost = newInterblockCost
+
+    val peerOpt = processingResult.transaction.source
+    peerOpt match {
+      case Some(peer) =>
+        val peerTxInfo = perPeerCost.getOrElse(peer, IncomingTxInfo.empty())
+        val newPeerCost = processingResult match {
+          case _: FailedTransaction => peerTxInfo.copy(invalidatedCost = peerTxInfo.invalidatedCost + cost)
+          case _: SuccessfulTransaction => peerTxInfo.copy(acceptedCost = peerTxInfo.acceptedCost + cost)
+          case _: DeclinedTransaction => peerTxInfo.copy(declinedCost = peerTxInfo.declinedCost + cost)
+        }
+        log.debug(s"Old peer ${peer.connectionId} cost info: ${peerTxInfo.totalCost}, " +
+          s"new: $newPeerCost, tx processing cache size: ${txProcessingCache.size}")
+        perPeerCost.put(peer, newPeerCost)
+      case _ => log.debug("No peer set, perPeerCost not updated.")
+    }
 
-    if (interblockCost.totalCost < MempoolCostPerBlock) {
+    if ((peerOpt.isDefined &&
+      perPeerCost.getOrElse(peerOpt.get, IncomingTxInfo.empty()).totalCost < MempoolPeerCostPerBlock) ||
+      interblockCost.totalCost < MempoolCostPerBlock) {
```

Comment:
Maybe a bit overkill here as the perPeerCost should always be set when the peer is defined.  But I figured it's better to continue gracefully in case of any edge cases.